### PR TITLE
feat(gateway): NPI Luhn + NUCC validation (#1150)

### LINKED ISSUE #1150: fhir-gateway: validate NPI (10-digit + Luhn) and NUCC (XXXXX0000X) shape before claiming us-core-practitioner conformance
## Context

Identified during review of PR #1141 (US Core Practitioner conformance, issue #947).

## Description

PR #1141 adds `users.npi` and `users.nucc_code` as nullable `text` columns and emits them into FHIR Practitioner resources. When BOTH are present the generator stamps `meta.profile = us-core-practitioner` as a conformance claim. However neither value is shape-validated anywhere in the write path or in the FHIR generator:

- **NPI**: should be exactly 10 digits and pass the [CMS NPI check-digit (modified Luhn) algorithm](https://www.cms.gov/Regulations-and-Guidance/Administrative-Simplification/NationalProvIdentStand/Downloads/NPIcheckdigit.pdf). The schema doc-comment says "10-digit string assigned by CMS" but nothing enforces it.
- **NUCC**: should match the `XXXXX0000X` taxonomy code pattern (5 alphanumerics + 4 zeros + 1 alpha). No regex check.

A malformed value populated by an ops backfill would currently sail through to a `meta.profile`-stamped Practitioner resource, which would fail any downstream US Core validator AND be a false conformance claim.

## Original Review Comment

> NPI shape validation — NOT validated (not 10-digit checked, not Luhn-checked). The schema doc-comment says "10-digit string assigned by CMS" but no enforcement at write or read time. Worth a defensive guard in the generator that withholds `meta.profile` (and possibly the NPI identifier entry) when shape doesn't match. Same applies to NUCC `XXXXX0000X` pattern.

## Acceptance Criteria

- [ ] Add a `NPI_REGEX` (`/^\d{10}$/`) + optional Luhn check helper in `services/fhir-gateway/src/generators/practitioner.ts` (or shared validators).
- [ ] Add a `NUCC_REGEX` (`/^[A-Z0-9]{5}0000[A-Z]$/i`) helper.
- [ ] In `toFhirPractitioner`, when `npi` or `nucc_code` fail their shape check: log a warning, emit the value WITHOUT `meta.profile` (treat as if NULL for conformance purposes), and emit the raw identifier/qualification anyway (caller can see it, just no conformance claim).
- [ ] Optional but ideal: enforce the same shape at write time in any auth-side mutation that updates `users.npi` / `users.nucc_code` (none exist today; add Zod validators when they do).
- [ ] Tests covering malformed NPI / NUCC: no `meta.profile`, fields still surface.

### PR DESCRIPTION
**Pull request by blamechris**
Closes #1150

## Summary

PR #1141 added `users.npi` and `users.nucc_code` columns and stamped `meta.profile = us-core-practitioner` whenever both were present — a real downstream conformance claim. Neither value was shape-validated, so a malformed value from an ops backfill would produce a false claim that HAPI / Inferno / Touchstone would later flag.

This PR adds shape validation at three layers (defence in depth):

### App-side validators (`services/fhir-gateway/src/generators/practitioner-validation.ts`)
- `isValidNpi(npi)` — 10 ASCII digits AND passes the CMS NPI check-digit algorithm (modified Luhn over the `80840`-prefixed string). Also rejects the all-zeros placeholder.
- `isValidNuccCode(code)` — matches the canonical XXXXX0000X taxonomy shape (5 alphanumeric + literal `0000` + single trailing letter), case-insensitive.

Both are pure functions; 13 dedicated unit tests cover the boundary cases (too short / too long / non-digit / Luhn failure / placeholder zeros / lowercase / missing 0000 segment / wrong terminal char).

### Practitioner generator (`services/fhir-gateway/src/generators/practitioner.ts`)
Tightened the conformance gate from "both present" to "both present AND shape-valid":
- Invalid NPI → drop the `us-npi` identifier entry; internal `/user-id` identifier still emitted.
- Invalid NUCC → drop the `qualification.code.coding` entry; text-only specialty still emitted.
- Either invalid → no `meta.profile` claim. Conservative: never assert a profile we can't back up.

4 new generator tests verify malformed-NPI, malformed-NUCC, both-malformed, and the valid-both happy path.

### DB CHECK constraints (`packages/db-schema/drizzle/0053_npi_nucc_check.sql`, idx 57)
- `npi` — `IS NULL OR ~ '^[0-9]{10}$'`
- `nucc_code` — `IS NULL OR ~ '^[A-Z0-9]{5}0000[A-Z]$'`

Postgres CHECK can't do Luhn without a plpgsql function, so the DB enforces shape and the app enforces Luhn. NULL remains valid (unbackfilled clinical-role rows).

## Notes

- Pre-existing test used `FAKE_NPI = "1234567890"` which is NOT Luhn-valid; updated to `1234567893` (textbook NPI Luhn example, verified manually). All 473 fhir-gateway tests pass.
- Migration idx 57 is the next slot after #1141's idx 56 (`0052_users_npi_nucc`).

## Test plan

- [x] `pnpm --filter @carebridge/fhir-gateway test -- practitioner` — 47/47 pass
- [x] `pnpm --filter @carebridge/fhir-gateway test` — 473/473 pass
- [x] `pnpm typecheck` — clean
- [x] `pnpm lint` — clean (existing unrelated warnings only)

### CODE REVIEW

**Comment by blamechris** [comment]
## Agent Review — PR #1157 (CareBridge Inspector)

**Verdict: APPROVE with minor follow-ups**

This PR cleanly closes the #1150 conformance gap: shape-validate `users.npi` (10 digits + CMS Luhn) and `users.nucc_code` (`XXXXX0000X`) before emitting them or claiming `us-core-practitioner` via `meta.profile`. Defense in depth (DB CHECK + app Luhn) is the right split — Postgres can't do Luhn, the app must.

### What I verified

- **NPI Luhn correctness**: Traced `isValidNpi("1234567893")` through the implementation manually. The CMS PDF textbook example (first 9 digits `123456789` → check digit `3`): prefix `80840`, right-to-left, double every second position, sum=70, mod 10=0 → valid. Implementation matches the CMS PDF exactly (rightmost = position 1, un-doubled).
- **NPI cross-check**: Ran the validator against 5 additional real NPIs from public NPPES sample data (`1679576722`, `1497758544`, `1003000126`, `1245319599`) — all accept. Rejection cases all reject (`1234567890`, `0000000000`, `9999999999`, `999999999X`).
- **Rejection coverage**: length 9/11, non-digit chars, leading/trailing whitespace, all-zeros placeholder — all explicitly tested.
- **Generator gating**: Traced all four branches (npi+nucc both valid → meta.profile; npi invalid → no us-npi identifier + no meta; nucc invalid → no qualification.coding + no meta; both invalid → neither + no meta). All correct, all tested.
- **DB migration**: idx 57 follows #1141's idx 56 correctly. Regex syntax valid Postgres (`~` POSIX, anchored). Constraint names follow `users_<col>_<purpose>_chk` convention.
- **Tests**: 17 new tests in practitioner-validation.test.ts (PR body says "13 dedicated unit tests" — minor undercount, not blocking). 473/473 fhir-gateway tests pass. typecheck + lint clean.
- **Fixed prior fixture**: `FAKE_NPI = "1234567890"` was indeed not Luhn-valid; the new `1234567893` is correct. Grep confirms no other stale NPI fixtures in the repo.
- **Write paths**: No `users.npi` / `users.nucc_code` write endpoints exist yet, so app-side validation only runs in the generator. When admin tools land, those will need to call `isValidNpi` + `isValidNuccCode` before insert (the PR description acknowledges this — issue #1150 acceptance criteria already calls it out as optional).
- **Zero attribution**: clean.

### Findings (follow-ups, not blocking)

**[High] NUCC case-sensitivity asymmetry (matches Copilot comments 1 + 2)**

`isValidNuccCode` uses `/i` flag (accepts lowercase), but the DB CHECK uses Postgres `~` (case-sensitive uppercase only). Also, `toFhirPractitioner` emits `user.nucc_code` as-is, so a lowercase value that passed the app validator would be exported in lowercase to the FHIR resource and stamped with `meta.profile = us-core-practitioner` — a false conformance claim, since canonical NUCC taxonomy codes are uppercase and Inferno/HAPI would flag lowercase.

In practice the DB CHECK blocks lowercase from being written, so this is currently latent. But the asymmetry is a footgun: drop the `/i` flag in `NUCC_SHAPE` and update the "accepts lowercase-cased input" test to assert lowercase is **rejected**, matching the DB and the canonical contract.

**[Med] Migration idempotency (matches Copilot comment 4)**

Other migrations adding CHECK constraints (e.g., `0026_family_access_constraints.sql`) prefix `ADD CONSTRAINT` with `DROP CONSTRAINT IF EXISTS`. Worth adopting here so dev/test re-runs don't fail.

**[Low / non-blocking] Pre-validate existing rows (Copilot comment 3)**

Copilot flagged that adding CHECK constraints could fail on historic malformed values. Checked: the columns were just added in #1141 (idx 56) and have no production writes, no seed data, and no write endpoints — there are no historic values. Migration is safe as-is. For future post-backfill schema changes, the `NOT VALID` + `VALIDATE CONSTRAINT` pattern is worth keeping in mind.

### Going to apply

Fix 1 (NUCC case-sensitivity) and Fix 2 (idempotency) in this PR. Reply to comment 3 with the rationale.

File: services/fhir-gateway/src/generators/practitioner-validation.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+ * NUCC taxonomy code shape: 5 alphanumeric chars + literal "0000" + a
+ * single trailing letter. The full canonical list is maintained at
+ * https://taxonomy.nucc.org/ and is too large / too volatile to bundle
+ * locally; we validate shape only and let the operator-driven backfill
+ * own correctness of the mapping.
+ *
+ * Case-insensitive; canonical codes from NUCC are uppercase. The DB
+ * CHECK constraint mirrors this shape (anchored, A-Z + digits only).
+ */
+const NUCC_SHAPE = /^[A-Z0-9]{5}0000[A-Z]$/i;
```
[the remark this answers is not in the document]
Comment (blamechris) [reply services/fhir-gateway/src/generators/practitioner-validation.ts]: Fixed in e5af4a7 — dropped the /i flag and made the regex case-sensitive. The asymmetry was a real footgun even though the DB CHECK was blocking lowercase writes: canonical NUCC codes are uppercase, and downstream FHIR validators compare exact case, so we want app + DB to agree. Test updated to assert lowercase and mixed-case are rejected.

File: packages/db-schema/drizzle/0053_npi_nucc_check.sql [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+-- NUCC code (the same NULL-tolerant contract documented on the
+-- columns in #947).
+
+ALTER TABLE users
+  ADD CONSTRAINT users_npi_shape_chk
+  CHECK (npi IS NULL OR npi ~ '^[0-9]{10}$');
+
+ALTER TABLE users
+  ADD CONSTRAINT users_nucc_code_shape_chk
+  CHECK (nucc_code IS NULL OR nucc_code ~ '^[A-Z0-9]{5}0000[A-Z]$');
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/db-schema/drizzle/0053_npi_nucc_check.sql:45]: Aligned by tightening the app side rather than loosening the DB — fixed in e5af4a7. The app validator is now case-sensitive (matching this CHECK) and the lowercase test was inverted to assert lowercase is rejected. Canonical NUCC is uppercase, so the DB is the right side of the contract to keep.

[the remark this answers is not in the document]
Comment (blamechris) [reply packages/db-schema/drizzle/0053_npi_nucc_check.sql:45]: Checked: the npi / nucc_code columns were just added in #1141 (idx 56) and have no production writes, no write endpoints, and no seed-data rows. There are no historic values to fail this migration. For future post-backfill schema changes, NOT VALID + VALIDATE CONSTRAINT is the right pattern. Filing a follow-up to add that to the docs/ops playbook.

File: packages/db-schema/drizzle/0053_npi_nucc_check.sql [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+-- Both constraints permit NULL because clinical-role rows that have
+-- not yet been backfilled with NPPES data legitimately have no NPI /
+-- NUCC code (the same NULL-tolerant contract documented on the
+-- columns in #947).
+
+ALTER TABLE users
+  ADD CONSTRAINT users_npi_shape_chk
+  CHECK (npi IS NULL OR npi ~ '^[0-9]{10}$');
+
+ALTER TABLE users
```
[the remark this answers is not in the document]
Comment (blamechris) [reply packages/db-schema/drizzle/0053_npi_nucc_check.sql:43]: Done in e5af4a7 — mirrored the DROP CONSTRAINT IF EXISTS pattern from 0026 and 0038.
